### PR TITLE
Develop Stream: fix warp dpp reduction

### PR DESCRIPTION
**Pull request by Beanavil**
1. **Fix intra-warp reduction**. By using rotations instead of shifts in [warp_reduce_dpp.hpp](https://github.com/ROCm/rocPRIM/blob/develop/rocprim/include/rocprim/warp/detail/warp_reduce_dpp.hpp#L66-L75) we make sure that no registers are left with invalid values. We thus substitute the `dpp_ctrl` values of `0x114` and `0x118` for `0x124` and `0x128`, respectively.
2. **Fix inter-warp reduction**. After performing the intra-warp reduction, a second round of reduction is done to reduce the warps' results. In this case, we still may have some threads reading invalid values because currently there is divergence in this reduction (only the first `warps_no_` threads will participate in the final reduction). So potentially many threads will be disabled and then the dpp instruction will resort to what `bound_ctrl` dictates just like described before. Thus, to ensure that all threads read and write valid values, we remove this divergence from [block_reduce_warp_reduce.hpp](https://github.com/ROCm/rocPRIM/blob/develop/rocprim/include/rocprim/block/detail/block_reduce_warp_reduce.hpp#L193).
3. Additionally, to prevent the same segfaults happening when the  `BlockSize` used is less than the hardware warp size, we added a fallback to the `warp_shuffle` in that case.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,12 @@
 
 Full documentation for rocPRIM is available at [https://rocm.docs.amd.com/projects/rocPRIM/en/latest/](https://rocm.docs.amd.com/projects/rocPRIM/en/latest/).
 
+## (Unreleased) rocPRIM 3.4.x for ROCm 6.4.x
+
+### Resolved issues
+
+* Fixed the warp-level reduction `rocprim::warp_reduce.reduce` DPP implementation to avoid undefined intermediate values during the reduction.
```
Comment (spolifroni-amd) [inline CHANGELOG.md]: If this has a github issue number, it can be added here.

**Review by spolifroni-amd — APPROVED** [review]
If there is a github issue number for the resolved issue, adding it to the changelog would be useful for users. Otherwise, all good.

File: CHANGELOG.md (hunk 1 above)
Comment (Beanavil) [reply CHANGELOG.md]: There is none afaik, this is for an internal ticket

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,12 @@
 
 Full documentation for rocPRIM is available at [https://rocm.docs.amd.com/projects/rocPRIM/en/latest/](https://rocm.docs.amd.com/projects/rocPRIM/en/latest/).
 
+## (Unreleased) rocPRIM 3.4.x for ROCm 6.4.x
```
Comment (stanleytsang-amd) [inline CHANGELOG.md]: Please change this to rocPRIM 3.4.0 for ROCm 6.4.0

**Review by stanleytsang-amd — CHANGES_REQUESTED** [review]
Hi @Beanavil, can you target this PR to the release-staging/rocm-rel-6.4 branch?  Thank you.

File: CHANGELOG.md (hunk 2 above)
Comment (Beanavil) [reply CHANGELOG.md]: Changed this and the target branch

**Review by NguyenNhuDi — APPROVED** [review]
lgtm

**Review by umfranzw — APPROVED** [review]
This change looks good to me.

**Pull request merged**